File: perl/pe015.pl
#!/usr/bin/perl -w
use strict;

# need to compute 2n!/((n!)*(n!))

my $limit = 20;
my ($numerator, $denominator);
for (my $i=0; $i < $limit; $i++) {
  push @$denominator, $i+1;
  push @$numerator, $i+$limit+1;
}

#  walk through lists and cast away multiples to reduce size of numbers from
#  factorials - this is incomplete factoring but sufficient for this problem
#  we could factor both completely if we wanted...

foreach my $i (@$denominator) {
  foreach my $j (@$numerator) {
    if (($i != 1) and (($j % $i) == 0)) {
      $j = $j / $i;
      $i = 1
    }
  }
}

my $num = 1;
my $denom = 1;
for (my $i = 0; $i < $limit; $i++) {
  $num *= $numerator->[$i];
  $denom *= $denominator->[$i];
}
    
print $num/$denom,"\n";

File: lua/pe015.lua
-- this computes projecteuler.net problem 015 

-- need to compute 2n!/((n!)*(n!))

local limit = 20
local numerator = {}
local denominator = {}

for i = 1, limit do
  denominator[i] = i
  numerator[i] = i+limit
end

--  walk through lists and cast away multiples to reduce size of numbers from
--  factorials - this is incomplete factoring but sufficient for this problem
--  we could factor both completely if we wanted...

for i = 1, limit do
  for j = 1, limit do
    local d = denominator[i]
    local n = numerator[j]
    if d ~= 1 and (n % d) == 0 then
        numerator[j] = math.floor(n / d)
        denominator[i] = 1
    end
  end
end

local num = 1
local denom = 1
for i = 1, limit do
  num = numerator[i] * num
  denom = denominator[i] * denom
end
      
print(math.floor(num/denom))
